framework/src/main/java/org/checkerframework/framework/source/CompositeChecker.java

```diff
@@ -0,0 +1,756 @@
+  // Not final because it is set in `init()`.
+  protected boolean warnUnneededSuppressions;
+
+  @Override
+  public void setRoot(CompilationUnitTree newRoot) {
+    super.setRoot(newRoot);
+    if (parentChecker == null) {
+      // Only clear the path cache if this is the main checker.
+      treePathCacher.clear();
+    }
+  }
+
+  /**
+   * Returns the set of subchecker classes on which this checker depends. Returns an empty set if
+   * this checker does not depend on any others.
+   *
+   * <p>Subclasses should override this method to specify subcheckers. If they do so, they should
+   * call the super implementation of this method and add dependencies to the returned set so that
+   * checkers required for reflection resolution are included if reflection resolution is requested.
+   *
+   * <p>Each subchecker of this checker may also depend on other checkers. If this checker and one
+   * of its subcheckers both depend on a third checker, that checker will only be instantiated once.
+   *
+   * <p>Though each checker is run on a whole compilation unit before the next checker is run, error
+   * and warning messages are collected and sorted based on the location in the source file before
+   * being printed. (See {@link #printOrStoreMessage(Diagnostic.Kind, String, Tree,
+   * CompilationUnitTree)}.)
+   *
+   * <p>WARNING: Circular dependencies are not supported nor do checkers verify that their
+   * dependencies are not circular. Make sure no circular dependencies are created when overriding
+   * this method. (In other words, if checker A depends on checker B, checker B cannot depend on
+   * checker A.)
+   *
+   * <p>This method is protected so it can be overridden, but it should only be called internally by
+   * the BaseTypeChecker.
+   *
+   * <p>The BaseTypeChecker will not modify the list returned by this method, but other clients do
+   * modify the list.
+   *
+   * @return the subchecker classes on which this checker depends; will be modified by callees
+   */
+  // This is never looked up in, but it is iterated over (and added to, which does a lookup).
+  protected Set<Class<? extends BaseTypeChecker>> getImmediateSubcheckerClasses() {
+    // This must return a modifiable set because clients modify it.
+    // Most checkers have 1 or fewer subcheckers.
+    LinkedHashSet<Class<? extends BaseTypeChecker>> result =
+        new LinkedHashSet<>(CollectionsPlume.mapCapacity(2));
+    if (shouldResolveReflection()) {
+      result.add(MethodValChecker.class);
+    }
+    return result;
+  }
+
+  /**
+   * Returns whether or not reflection should be resolved.
+   *
+   * @return true if reflection should be resolved
+   */
+  public boolean shouldResolveReflection() {
+    return hasOptionNoSubcheckers("resolveReflection");
+  }
+
+  /** An array containing just {@code BaseTypeChecker.class}. */
+  protected static Class<?>[] baseTypeCheckerClassArray = new Class<?>[] {BaseTypeChecker.class};
```

Comment:
It's very strange that this field needs to be in `CompositeChecker` instead of `BaseTypeChecker`. This is definitely a code smell, so I'd like to understand why this field is here.